pkg/tools/clientmap/refcount.go

```diff
@@ -17,73 +17,114 @@
+// loaded result is true if the value was loaded, false if stored.
+// store  -->  count = 1
+// load   -->  count += 1
+func (m *RefcountMap) LoadOrStore(key string, value networkservice.NetworkServiceClient) (_ networkservice.NetworkServiceClient, loaded bool) {
+	var raw interface{}
+	raw, loaded = m.m.LoadOrStore(key, &entry{
+		count: 1,
+		value: value,
+	})
+	entry := raw.(*entry)
+	if !loaded {
+		return entry.value, false
+	}
 
-// Store sets the value for a key.
-// Increments the refcount
-func (r *RefcountMap) Store(key string, value networkservice.NetworkServiceClient) {
-	client := &refcountClient{client: value}
-	r.Map.Store(key, client)
-	atomic.AddInt32(&client.count, 1)
-}
+	entry.lock.Lock()
 
-// LoadOrStore returns the existing value for the key if present. Otherwise, it stores and returns the given value. The loaded result is true if the value was loaded, false if stored.
-// Increments the refcount.
-func (r *RefcountMap) LoadOrStore(key string, value networkservice.NetworkServiceClient) (networkservice.NetworkServiceClient, bool) {
-	client := &refcountClient{client: value}
-	rv, loaded := r.Map.LoadOrStore(key, client)
-	if client, ok := rv.(*refcountClient); ok {
-		atomic.AddInt32(&client.count, 1)
-		return client.client, loaded
+	if entry.count == 0 {
+		entry.lock.Unlock()
+		return m.LoadOrStore(key, value)
 	}
-	return rv, loaded
+	entry.count++
+
+	entry.lock.Unlock()
+
+	return entry.value, true
 }
 
-// Load returns the value stored in the map for a key, or nil if no value is present. The ok result indicates whether value was found in the map.
-// Increments refcount.
-func (r *RefcountMap) Load(key string) (networkservice.NetworkServiceClient, bool) {
-	rv, loaded := r.Map.Load(key)
-	if client, ok := rv.(*refcountClient); ok {
-		atomic.AddInt32(&client.count, 1)
-		return client.client, loaded
+// Load returns the value stored in the map for a key, or nil if no value is present. The loaded result indicates
+// whether value was found in the map.
+// count += 1
+func (m *RefcountMap) Load(key string) (_ networkservice.NetworkServiceClient, loaded bool) {
+	var raw interface{}
+	raw, loaded = m.m.Load(key)
+	if !loaded {
+		return nil, false
+	}
+	entry := raw.(*entry)
+
+	entry.lock.Lock()
+	defer entry.lock.Unlock()
+
+	if entry.count == 0 {
+		return nil, false
 	}
-	return rv, loaded
+	entry.count++
+
+	return entry.value, true
 }
 
-// Delete decrements the refcount and deletes the value for a key if refcount is zero. Returns true if value is no (more) present.
-func (r *RefcountMap) Delete(key string) bool {
-	rv, loaded := r.Map.Load(key)
+// LoadAndDelete is trying to delete the value for a key, returning the previous value if any. The loaded result
+// reports whether the key was present.
+// count == 1  -->  delete
+// count > 1   -->  count -= 1
+func (m *RefcountMap) LoadAndDelete(key string) (_ networkservice.NetworkServiceClient, loaded, deleted bool) {
```

Comment:
Not actually, because not every `Delete`, `LoadAndDelete` deletes a value for the key as it is in the `sync.Map`. So we probably want to know is the value has been deleted for real. And so we need to add additional `bool` return parameter for both `Delete` and `LoadAndDelete`.

Reply:
As I can see proposed api is not using yet. So I think we should follow sync.Map LoadAndDelete signature.

Reply:
Here it is used :)
https://github.com/networkservicemesh/sdk/blob/98deb86e4ee85abd4622882171b79433be93c335/pkg/networkservice/common/connect/server.go#L142